compiler/paths.jou

```diff
@@ -98,97 +216,177 @@ def find_stdlib() -> byte*:
 
-    subfolder = get_filename_without_jou_suffix(command_line_args.infile)
 
-    result: byte* = malloc(strlen(infile_folder) + strlen(subfolder) + strlen(filename) + 100)
-    assert result != NULL
+# When compiling file foo.jou, we use jou_compiled/foo/ subdirectory.
+#
+# This is currently needed to avoid race conditions when running multiple files
+# at the same time. Feel free to clean this up if you implement locking.
+# See also: https://github.com/Akuli/jou/issues/743
+def get_path_to_jou_compiled_subfolder() -> byte*:
+    # Place jou_compiled into the same directory with the main file we are compiling.
+    # No strong reason for doing this. Feel free to change this if it causes problems.
+    infile = strdup(command_line_args.infile)
+    jou_compiled: byte*
+    asprintf(&jou_compiled, "%s/jou_compiled", dirname(infile))
+    free(infile)
+
+    simple_mkdir(jou_compiled)
+    write_gitignore(jou_compiled)
+
+    name = skip_slashes(command_line_args.infile)
+    name_len = strlen(name)
+    if name_len > 4 and ends_with(name, ".jou"):
+        name_len -= 4
+
+    subfolder: byte*
+    asprintf(&subfolder, "%s/%.*s", jou_compiled, name_len as int, name)
+    simple_mkdir(subfolder)
+
+    free(jou_compiled)
+    return subfolder
+
+
+# Deletes files in a directory whose name ends with the given suffix.
+def delete_by_suffix(dir: byte*, suffix: byte*) -> None:
+    list = listdir(dir)
+
+    for p = list.ptr; p < list.end(); p++:
+        if ends_with(*p, suffix):
+            full_path: byte*
+            asprintf(&full_path, "%s/%s", dir, *p)
 
-    sprintf(result, "%s/jou_compiled", infile_folder)
-    mkdir_exist_ok(result)
-    write_gitignore(result)
+            if WINDOWS:
+                ret = _unlink(full_path)
+            else:
+                ret = unlink(full_path)
 
-    sprintf(result, "%s/jou_compiled/%s", infile_folder, subfolder)
-    mkdir_exist_ok(result)
+            if ret == 0 and command_line_args.verbosity >= 1:
+                printf("Deleted %s\n", full_path)
 
-    sprintf(result, "%s/jou_compiled/%s/%s", infile_folder, subfolder, filename)
+            free(full_path)
+
+    free(list.ptr)
+
+
+def get_simple_name(jou_file_path: byte*) -> byte[50]:
+    jou_file_path = skip_slashes(jou_file_path)
+
+    result: byte[50]
+    memset(result, 0, sizeof(result))
+    result_len = 0
+
+    while result_len + 1 < sizeof(result):
+        # We have room to add a character
+
+        if jou_file_path[0] == '\0':
+            # No more characters to add
+            break
+
+        if result_len > 0 and strcmp(jou_file_path, ".jou") == 0:
+            # do not add ".jou" suffix
+            break
+
+        c = *jou_file_path++
+        if c == '.':
+            c = '_'
+        result[result_len++] = c
 
-    free(infile_folder)
-    free(subfolder)
     return result
 
 
+def calculate_file_path_hash(jou_file_path: byte*) -> Hash:
+    h = new_hash()
+    h.add_bytes(jou_file_path, strlen(jou_file_path))
```

Comment:
It would be better to include optimization level in this hash, so that -O0, -O1, -O2 and -O3 versions of the object file can coexist and not get deleted when recompiling with more/less optimizations.